### PR TITLE
Update `GET /followers` API endpoint

### PR DESCRIPTION
**Pull request by citizen428**
## What type of PR is this? (check all applicable)

- [X] Feature

## Description

This PR makes 2 changes to the `GET /followers` API endpoint:

1. The `created_at` value of the follow is now included in the API response.
2. A sort order can be specified (before it was hard-coded to `created_at: :desc`).

Note: I used `sort` as the parameter name because we use the same name here: https://github.com/forem/forem/pull/13137/files

## Related Tickets & Documents

https://github.com/forem/internalEngineering/issues/404

## QA Instructions, Screenshots, Recordings

1. Verify that the `created_at` timestamp is part of the API response.
2. Verify that sorting works.

### UI accessibility concerns?

None, API only.

## Added tests?

- [X] Yes

## [Forem core team only] How will this change be communicated?

- [X] I will share this change internally with the appropriate teams

### CODE REVIEW

File: docs/api_v0.yml [hunk 1]
```diff
@@ -695,6 +696,8 @@ components:
       properties:
         type_of:
           type: string
+        created_at:
+          type: string
```
Comment (rhymes) [inline docs/api_v0.yml:700]: we should add `format: date-time` if possible

File: app/views/api/v0/followers/users.json.jbuilder [hunk 2]
```diff
@@ -1,6 +1,7 @@
 json.array! @follows do |follow|
-  json.type_of "user_follower"
-  json.id      follow.id
+  json.type_of    "user_follower"
+  json.id         follow.id
+  json.created_at follow.created_at
```
Comment (rhymes) [inline app/views/api/v0/followers/users.json.jbuilder:4]: please stick to ISO/RFC3339 format, see other endpoints. You can use 

https://github.com/forem/forem/blob/14416f79cc02f002e34f53339ed28a27adee1a48/app/helpers/articles_helper.rb#L56-L58

File: docs/api_v0.yml [hunk 3]
```diff
@@ -1552,12 +1555,14 @@ components:
       value:
         - type_of: user_follower
           id: 12
+          created_at: "2021-04-07 08:29:42 UTC"
```
Comment (rhymes) [inline docs/api_v0.yml:1558]: the dates need to be updated with the ISO format once changed

**Review by rhymes — COMMENTED** [review]
A tiny difference that isn't clear by the the diff in the [previous PR](https://github.com/forem/forem/pull/13137/files) for the articles endpoint.

In our conversation I suggested using `sort=created_at:desc`, that's because you can both specify which parameter the endpoint sorts against and the order.

We didn't need it in that PR because the public API is `/articles/latest` which defaults to `published_at: :desc` but since in this case we're going to introduce an entirely new concept to the client, I think we should decide the interface once and for all, so we don't have to come up with `sort_by` and `sort_order` down the road. `sort=asc` doesn't include which field in case there are many.

I think we have two strategies before us:

- JSON API's [fetching and sorting](https://jsonapi.org/format/#fetching-sorting) which uses: `sort=-created_at,title` where `-created_at` means `created_at: :desc` and `title` means `title: :asc`
- Another strategy [I found online](https://www.taniarascia.com/rest-api-sorting-filtering-pagination/#multiple-columns) which uses: `sort=created_at:desc,title:asc`

I'm good either way, but I think we should be explicit instead of using `sort=asc`.

File: app/controllers/api/v0/followers_controller.rb [hunk 4]
```diff
@@ -5,15 +5,15 @@ class FollowersController < ApiController
...
+        id follower_id follower_type created_at
       ].freeze
       private_constant :USERS_ATTRIBUTES_FOR_SERIALIZATION
 
       def users
         @follows = Follow.followable_user(@user.id)
           .includes(:follower)
           .select(USERS_ATTRIBUTES_FOR_SERIALIZATION)
-          .order(created_at: :desc)
+          .order(created_at: sort_param)
```
Comment (jgaskins) [inline app/controllers/api/v0/followers_controller.rb:16]: It looks like `created_at` is the sort param here, and this is actually the sort direction.

```suggestion
          .order(created_at: sort_direction)
```

File: app/controllers/api/v0/followers_controller.rb [hunk 5]
```diff
@@ -24,6 +24,12 @@ def limit_per_page(default:, max:)
         per_page = (params[:per_page] || default).to_i
         @follows_limit = [per_page, max].min
       end
+
+      def sort_param
```
Comment (jgaskins) [inline app/controllers/api/v0/followers_controller.rb:28]: To match the above

```suggestion
      def sort_direction
```

File: app/controllers/api/v0/followers_controller.rb [hunk 6]
```diff
@@ -24,6 +24,12 @@ def limit_per_page(default:, max:)
...
+        return :asc if params[:sort] == "asc"
```
Comment (jgaskins) [inline app/controllers/api/v0/followers_controller.rb:29]: Looking at it from the perspective of a consumer of this API, what does it mean to sort followers in "ascending" order?

File: app/controllers/api/v0/followers_controller.rb (hunk 4 above)
Comment (citizen428) [reply app/controllers/api/v0/followers_controller.rb:16]: Well, it is the param called `sort` but that doesn't change the fact that it was a horrible method name. 😆 Thanks!

File: app/controllers/api/v0/followers_controller.rb (hunk 6 above)
Comment (citizen428) [reply app/controllers/api/v0/followers_controller.rb:29]: This ties in with @rhymes comment above re sort params and the future direction of sorting in the API: https://github.com/forem/forem/pull/13287#pullrequestreview-629799466

**Comment by citizen428** [comment]
@rhymes @jgaskins I now added a concern to deal with JSON API style sort parameters and updated the code to use it.

**Review by rhymes — COMMENTED** [review]
Looks good! Love that we can also specify which fields are allowed so we don't get surprises on sorting on fields that are private or have no indexes <3 

Left a note about updating the API documentation

File: docs/api_v0.yml [hunk 7]
```diff
@@ -2709,6 +2717,16 @@ paths:
...
+        - in: query
+          name: sort
+          schema:
+            type: string
+            default: desc
+          required: false
+          description: |
+            Specifies the sort order for the `created_at` param of the follow
+            relationship. It defaults to `desc`, to get reverse order `asc` can
+            be used.
```
[the remark this answers is not in the document]
Comment (citizen428) [reply docs/api_v0.yml]: Updated, thanks!

**Comment by citizen428** [comment]
Sorry for the force push I resolved merge conflicts but did so by rebasing against `master`, not merging it back here.

File: Gemfile.lock [hunk 8]
```diff
@@ -86,7 +86,7 @@ GEM
       activerecord (>= 5.0, < 6.2)
```
Comment (rhymes) [inline Gemfile.lock:86]: I'm not sure this file and all the other gems should be here

**Review by rhymes — COMMENTED** [review]
Works great! Thanks for the changes. We just need to remove the `Gemfile.lock` and the gem changes. Something must have happened during the rebase...

File: Gemfile.lock (hunk 8 above)
Comment (citizen428) [reply Gemfile.lock]: They shouldn't, all fixed ✔️

**Review by rhymes — APPROVED** [review]
LGTM!

**Review by jacobherrington — APPROVED** [review]
LGTM and works as expected 👍

**Comment by bencgreenberg** [comment]
This is awesome! Thank you amazing Forem team!

**Pull request merged**